packages/hoppscotch-cli/src/utils/mutators.ts

```diff
@@ -39,49 +34,46 @@ export const parseErrorMessage = (e: unknown) => {
   return msg.replace(/\n+$|\s{2,}/g, "").trim();
 };
 
+export async function readJsonFile(path: string) {
+  if(!path.endsWith('.json')) {
+    throw error({ code: "INVALID_FILE_TYPE", data: path })
+  }
+  try {
+    await fs.access(path)
+  } catch (e) {
+    throw error({ code: "FILE_NOT_FOUND", path: path })
+  }
+  try { 
+    return JSON.parse((await fs.readFile(path)).toString())
+  } catch(e) {
+    throw error({ code: "UNKNOWN_ERROR", data: e })
+  }
+}
+
 /**
  * Parses collection json file for given path:context.path, and validates
  * the parsed collectiona array.
  * @param path Collection json file path.
  * @returns For successful parsing we get array of HoppCollection<HoppRESTRequest>,
  */
-export const parseCollectionData = (
+export async function parseCollectionData(
   path: string
-): TE.TaskEither<HoppCLIError, HoppCollection<HoppRESTRequest>[]> =>
-  pipe(
-    TE.of(path),
-
-    // Checking if given file path exists or not.
-    TE.chain(checkFile),
-
-    // Trying to read give collection json path.
-    TE.chainW((checkedPath) =>
-      TE.tryCatch(
-        () => fs.readFile(checkedPath),
-        (reason) => error({ code: "UNKNOWN_ERROR", data: E.toError(reason) })
-      )
-    ),
-
-    // Checking if parsed file data is array.
-    TE.chainEitherKW((data) =>
-      pipe(
-        data.toString(),
-        J.parse,
-        E.map((jsonData) => (Array.isArray(jsonData) ? jsonData : [jsonData])),
-        E.mapLeft((e) =>
-          error({ code: "MALFORMED_COLLECTION", path, data: E.toError(e) })
-        )
-      )
-    ),
-
-    // Validating collections to be HoppRESTCollection.
-    TE.chainW(
-      TE.fromPredicate(A.every(isRESTCollection), () =>
-        error({
-          code: "MALFORMED_COLLECTION",
-          path,
-          data: "Please check the collection data.",
-        })
-      )
-    )
-  );
+): Promise<HoppCollection<HoppRESTRequest>[]> {
+  let contents = await readJsonFile(path)
+  if (!Array.isArray(contents)) {
+    contents = [contents]
+  }
+  const out:HoppCollection<HoppRESTRequest>[] = []
+  for(const maybeValidCollection of contents) {
+    if(isRESTCollection(maybeValidCollection)) {
+      out.push(maybeValidCollection)
+    } else {
+      throw error({
+              code: "MALFORMED_COLLECTION",
+              path,
+              data: "Please check the collection data.",
+            })
+    }
+  }
+  return out
```

Comment:
It looks like there is no point in filling up an array again as if this function ran without an error, `out` will basically be equivalent to `contents`. Hence we can just do this.

```suggestion
  if (contents.some((x) => !isRESTCollection(x))) {
    throw error({
      code: "MALFORMED_COLLECTION",
      path,
      data: "Please check the collection data.",
    })
  }

  return contents
```

Reply:
Validating original array and returning at https://github.com/hoppscotch/hoppscotch/pull/2907/commits/e94fb04215071caf6ab5b1910d6df0a6a63294a1#diff-fccb5e9c3967b37fa6f0242a9bdddcfcad1f7e7baf5f4508f95730ccf78bdf97L78